torch/_decomp/__init__.py

```diff
@@ -250,13 +253,195 @@ def remove_decompositions(
+# decomp part. (https://github.com/pytorch/pytorch/issues/129431)
+def _check_valid_to_preserve(op_overload):
+    if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+        return False
+    if op_overload in FunctionalTensor.metadata_fns:
+        return False
+
+    alias_info = len(
+        [i for i in op_overload._schema.arguments if i.alias_info is not None]
+    )
+
+    is_mutating_or_aliasing = alias_info != 0 or op_overload._schema.is_mutable
+
+    if is_mutating_or_aliasing:
+        return False
+
+    if not torch._C._dispatch_has_kernel(op_overload.name()):
+        return False
+
+    return True
+
+
+def _is_cia_op(op: "OpOverload") -> bool:
+    return (
+        torch._C._dispatch_has_kernel_for_dispatch_key(
+            op.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        )
+        or torch._C.DispatchKey.CompositeImplicitAutograd in op.py_kernels
+    )
+
+
+_ALL_VALID_CIA_OPS_IN_EXPORT: Optional[frozenset["OpOverload"]] = None
+
+
+def _collect_all_valid_cia_ops() -> frozenset["OpOverload"]:
+    """
+    This is an util function that gets the all CIA functional ops.
+
+    The algorithm is in 2 steps:
+      1. We first query C++ dispatcher to get the list of CIA ops
+         and then we call getattr on torch.ops.aten to lazily populate
+         them.
+
+      2. Sometimes, handful of ops have CIA registered in python dispatcher
+         but not on the C++ side, these can't be caught at the first step.
+         So we walk again to get the final list.
+
+    We need a frozenset as output because the set can be modified by user.
+    """
+
+    # This is an expensive operation, so we cache the result
+    global _ALL_VALID_CIA_OPS_IN_EXPORT
+    if _ALL_VALID_CIA_OPS_IN_EXPORT is not None:
+        return _ALL_VALID_CIA_OPS_IN_EXPORT
+
+    # First step to lazily populate torch.ops.aten
+    cia_ops = torch._C._dispatch_get_registrations_for_dispatch_key(
+        "CompositeImplicitAutograd"
+    )
+    # Ignore quantized namespace ops
+    cia_ops = [name[6:] for name in cia_ops if name.startswith("aten::")]
+    # Materialize all CIA ops first
+    for op in cia_ops:
+        split_list = op.split(".")
+        # Sometime overload could be missing
+        assert len(split_list) == 1 or len(split_list) == 2
+        op_name = split_list[0]
+        op_overload_name = "default"
+        if len(split_list) == 2:
+            op_overload_name = split_list[1]
+
+        _ = getattr(getattr(torch.ops.aten, op_name), op_overload_name)
+
+    # Second step to finally compile the list of all valid ops
+    cia_ops = set()
+    for op in torch.ops.aten:
+        op_packet = getattr(torch.ops.aten, op)
+        for overload in op_packet.overloads():
+            op_overload = getattr(op_packet, overload)
+            if _check_valid_to_preserve(op_overload) and _is_cia_op(op_overload):
+                cia_ops.add(op_overload)
+    _ALL_VALID_CIA_OPS_IN_EXPORT = frozenset(cia_ops)
```

Comment:
cc @bdhirsh 

I made a small benchmark to see how much this caching helps. It reduces 1 min to 45 second which i guess is worth it. Without frozenset, it is 23 seconds but it is very not safe.

Reply:
The script i used was:
```
import torch 

class ConvBatchnorm(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 3, 1, 1)
        self.bn = torch.nn.BatchNorm2d(3)

    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        return (x,)

mod = ConvBatchnorm()
inp = torch.randn(1, 1, 3, 3)

ep = torch.export.export_for_training(mod, (inp,))
# Lower to core aten inference IR, but keep conv2d
decomp_table = torch._decomp.core_aten_decompositions()


for _ in range(100):
    def my_awesome_custom_conv2d(x, weight, bias, stride=[1, 1], padding=[0, 0], dilation=[1, 1], groups=1):
        return 2 * torch.ops.aten.convolution(x, weight, bias, stride, padding, dilation, False, [0, 0], groups)

    decomp_table[torch.ops.aten.conv2d.default] = my_awesome_custom_conv2d
    ep_for_inference = ep.run_decompositions(decomp_table)

    print(ep_for_inference)
```

Reply:
hmm, two questions:

(1) can you just stick an `@lru_cache` on top of `_collect_all_valid_cia_ops`?

(2) i'm a bit confused by the benchmark - if you cache the result of this function, then won't the first export call that you do be slower, but all future calls be faster? (also - is looping over all of `torch.ops.aten` taking 30+ seconds??)